src/objects.c

```diff
@@ -387,6 +387,36 @@ PgUser *add_user(const char *name, const char *passwd)
 	return user;
 }
 
+/* find original or preconfigured user given the login user */
+PgUser *find_original_user(PgUser *login_user)
+{
+	/*
+	 * The original user will either be the current login user, a preconfigured user via
+	 * ini file [users] section, a preconfigured user via global auth_file, a PAM user,
+	 * a user attached to a database via auth_query, a forced user with user= entry via
+	 * ini file [databases] section, or not exist yet if the user has never connected.
+	 */
+	PgUser *original_user = find_user(login_user->name);
+	char *tmp_passwd;
+
+	/* add new login user to reload their configurations at runtime */
+	if (original_user == NULL) {
+		put_in_order(&login_user->head, &user_list, cmp_user);
+		aatree_insert(&user_tree, (uintptr_t)login_user->name, &login_user->tree_node);
+		return login_user;
+	}
+
+	/* original user is no longer preconfigured and won't send auth queries on future login */
+	if (original_user->is_preconfigured)
+		original_user->is_preconfigured = false;
+
+	tmp_passwd = strdup(login_user->passwd);
+	safe_strcpy(original_user->passwd, tmp_passwd, sizeof(original_user->passwd));
```

Comment:
Copy over password of new login user to account for a single user that may have different passwords for multiple databases.

Reply:
So you are copying every time in case a user has a different password in different databases? I think even with this code that could lead to problems, because the `original_user` is shared between clients of different databases. So maybe we should just document that a user should not have different passwords for different databases when using `auth_query`/`auth_user`.

Reply:
Updated the PR. `find_original_user` no longer exists. To support a user logging into different databases with different passwords (usually when databases have different hosts), each `PgDatabase*` now stores a list of user passwords.

```
struct PgUserPassword {
	struct AANode tree_node;	/* used to attach user password to tree */
	char username[MAX_USERNAME];
	char passwd[MAX_PASSWORD];
};
```

Semantically, this should be clearer since it models reality, where each database may have different passwords per user. `PgUser*` still retains a password field to store a password from `auth_file` which will be used if it exists by default.

Reply:
This change seems to somehow have been reverted on your branch. Possibly due to a rebase.